open_rarity/rarity_ranker.py

```diff
@@ -20,35 +28,76 @@ def rank_tokens(*, token_id_to_scores: dict[str, float]) -> dict[str, int]:
 
         Parameters
         ----------
-        token_ids_to_score : dict[str, float]
-            A dictionary of token ids (this can be whatever id you want, as long
-            as its unique) to the rarity score.
+        collection : Collection
+            Collection object with populated tokens
+        scorer: Scorer
+            Scorer instance
+
+        Returns
+        -------
+        Collection
+            Collection with populated rarity ranks and scores
+        """
+
+        if (
+            collection is None
+            or collection.tokens is None
+            or len(collection.tokens) == 0
+        ):
+            return collection
+
+        tokens = collection.tokens
+        scores: list[float] = scorer.score_tokens(collection, tokens=tokens)
+
+        # fail ranking if dimension of scores doesn't match dimension of tokens
+        assert len(tokens) == len(scores)
+
+        # augment collection tokes with score information
+        for idx, token in enumerate(tokens):
+            token.token_rarity = TokenRarity(score=scores[idx])
+
+        # assign sorted tokens list to the collection object
+        collection._tokens = RarityRanker.rank_tokens(collection.tokens)
```

Comment:
Don't think it's best practice to be modifying inputs or private variables like this - The reason we initially made tokens a private variable is to ensure no re-setting.

Reply:
The main reason to update here is to preserve the sort order - we can remove this from the scope and return non-sorted list.

Reply:
Ok after a bit of thinking i inversed the relationship and the need to mutation disappeared.

Reply:
fixed